File: rust/year_2018/src/day_02.rs
use itertools::Itertools;

pub fn part_1(input: &str) -> usize {
    let counts = input
        .lines()
        .flat_map(|line| line.chars().counts().into_values().unique())
        .counts();
    [2, 3]
        .iter()
        .map(|count| counts.get(count).unwrap_or(&0))
        .product()
}

pub fn part_2(input: &str) -> String {
    input
        .lines()
        .flat_map(|line| {
            (0..line.len()).map(|i| {
                line.chars()
                    .enumerate()
                    .map(|(j, char)| (i != j).then_some(char))
                    .collect_vec()
            })
        })
        .duplicates()
        .next()
        .expect("some duplicate")
        .into_iter()
        .flatten()
        .collect::<String>()
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const EXAMPLE_1: &str = "abcdef
bababc
abbcde
abcccd
aabcdd
abcdee
ababab";

    const INPUT: &str = include_str!("../test_data/day_02.txt");

    #[test_case(EXAMPLE_1 => 12)]
    #[test_case(INPUT => 5681)]
    fn part_1(input: &str) -> usize {
        super::part_1(input)
    }

    const EXAMPLE_2: &str = "abcde
fghij
klmno
pqrst
fguij
axcye
wvxyz";

    #[test_case(EXAMPLE_2 => "fgij")]
    #[test_case(INPUT => "uqyoeizfvmbistpkgnocjtwld")]
    fn part_2(input: &str) -> String {
        super::part_2(input)
    }
}

File: python/year_2018/day_02.py
"""
https://adventofcode.com/2018/day/2
"""

from collections import Counter
from pathlib import Path

from iteration_utilities import Seen


def part_1(text):
    def checksum(line):
        frequencies = set(Counter(line).values())
        return complex(2 in frequencies, 3 in frequencies)

    total = sum(map(checksum, text.splitlines()))
    return int(total.imag * total.real)


def part_2(text):
    lines = text.strip().split()
    length = len(lines[0])
    commons = [Seen() for _ in range(length)]

    for line in lines:
        for i, common in enumerate(commons):
            string = f"{line[:i]}{line[i + 1:]}"
            if common.contains_add(string):
                return string


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_02.txt").read_text()


EXAMPLE_1 = """abcdef
bababc
abbcde
abcccd
aabcdd
abcdee
ababab
"""

EXAMPLE_2 = """abcde
fghij
klmno
pqrst
fguij
axcye
wvxyz
"""


def test_part_1_input():
    assert part_1(_read_input()) == 5681


def test_part_1_example_1():
    assert part_1(EXAMPLE_1) == 12


def test_part_1_example_2():
    assert part_1(EXAMPLE_2) == 0


def test_part_2_input():
    expected = "uqyoeizfvmbistpkgnocjtwld"
    assert part_2(_read_input()) == expected


def test_part_2_example_1():
    assert part_2(EXAMPLE_1) == "abcde"


def test_part_2_example_2():
    assert part_2(EXAMPLE_2) == "fgij"
